src/main/java/mate/academy/springboot/datajpa/repository/ProductRepositoryCustomImpl.java

```diff
@@ -0,0 +1,35 @@
+package mate.academy.springboot.datajpa.repository;
+
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.List;
+import java.util.Map;
+import mate.academy.springboot.datajpa.model.Product;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class ProductRepositoryCustomImpl implements ProductRepositoryCustom {
+    @PersistenceContext
+    private EntityManager em;
+
+    @Override
+    public List<Product> findAll(Map<String, String> params) {
+        CriteriaBuilder cb = em.getCriteriaBuilder();
```

Comment:
```suggestion
@Repository
public class ProductSpecification {
    public Specification<Product> extractSpecification(Map<String, String> params) {
        return (root, query, cb) -> {
             root.fetch(...)
             List<Predicate> predicates ...
             
             ...
             
             return cb.and(predicates.toArray(new Predicate[] {})
```
Потім можеш використати обєкт з цього метода в метода findAll(Specification)
замість того аби створювати імплентацію, можна створити свою специфікацію